test/distributed/_spmd/test_transformation.py

```diff
@@ -0,0 +1,142 @@
+        ddp_model = DDP(deepcopy(model), device_ids=[dist.get_rank()])
+        optim = torch.optim.Adam(
+            model.parameters(), lr=0.01, foreach=True, capturable=True
+        )
+        ddp_optim = torch.optim.Adam(
+            ddp_model.parameters(), lr=0.01, foreach=True, capturable=True
+        )
+        batch = torch.randn(batch_size, dim).cuda()
+
+        # materialize optimizer states
+        out = model(batch)
+        out.sum().backward()
+        optim.step()
+        optim.zero_grad()
+
+        ddp_out = ddp_model(batch)
+        ddp_out.sum().backward()
+        ddp_optim.step()
+        ddp_optim.zero_grad()
+
+        self.assertEqual(ddp_out, out)
+        self.assertEqual(list(ddp_model.parameters()), list(model.parameters()))
+        return model, optim, ddp_model, ddp_optim
+
+    def _test_tran_step_with_ddp_without_optim_step(
+        self, train_step, num_iters, batch_size, layers, dim
+    ):
+        def _ddp_train_step(model, optim, batch):
+            model(batch).sum().backward()
+            return [p.grad for p in model.parameters()]
+
+        model, optim, ddp_model, ddp_optim = self._init(batch_size, layers, dim)
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
+            ddp_out = _ddp_train_step(ddp_model, ddp_optim, batch)
+            for g1, g2 in zip(out, ddp_out):
+                self.assertEqual(g1 / self.world_size, g2)
+            optim.zero_grad()
+            ddp_optim.zero_grad()
+
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_basic_transformation(self):
+        batch_size = 100
+        layers = 10
+        dim = 100
+        num_iters = 5
+
+        @compile(gm_transformation=GraphModuleTransformation(num_iters=num_iters))
+        def train_step(model, optim, batch):
+            model(batch).sum().backward()
+            return [p.grad for p in model.parameters()]
+
+        self._test_tran_step_with_ddp_without_optim_step(
+            train_step, num_iters, batch_size, layers, dim
+        )
+
+    @unittest.skipIf(not has_triton(), "Inductor+gpu needs triton and recent GPU arch")
+    @skip_if_lt_x_gpu(2)
+    @with_comms
+    def test_inductor(self):
+        batch_size = 100
+        layers = 10
+        dim = 100
+        num_iters = 5
+
+        @compile(
+            gm_transformation=GraphModuleTransformation(
+                num_iters=num_iters, enable_inductor=True
+            )
+        )
+        def train_step(model, optim, batch):
+            model(batch).sum().backward()
+            return [p.grad for p in model.parameters()]
+
+        model, optim, _, _ = self._init(batch_size, layers, dim)
+        for _ in range(num_iters):
+            batch = torch.randn(batch_size, dim).cuda()
+            out = train_step(model, optim, batch)
+
+            # TODO: using _test_tran_step_with_ddp_without_optim_step()
```

Comment:
Hey @fegin, do you plan to complete this test before landing, or is this for followup PRs?

Reply:
Done in the latest version.